Source/WebCore/layout/formattingContexts/grid/GridLayoutUtils.cpp

```diff
@@ -43,11 +43,59 @@ LayoutUnit computeGapValue(const Style::GapGutter& gap)
 {
+    auto columnsSize = [&] {
+        auto columnsStartPosition = computeGridLinePosition(placedGridItem.columnStartLine(), usedColumnSizes, columnsGap);
+        auto columnsEndPosition = computeGridLinePosition(placedGridItem.columnEndLine(), usedColumnSizes, columnsGap);
+        ASSERT(columnsEndPosition >= columnsStartPosition);
+        return columnsEndPosition - columnsStartPosition;
+    };
+
     auto& inlineAxisSizes = placedGridItem.inlineAxisSizes();
-    if (auto fixedInlineSize = inlineAxisSizes.preferredSize.tryFixed())
-        return LayoutUnit { fixedInlineSize->resolveZoom(placedGridItem.usedZoom()) };
+    ASSERT(inlineAxisSizes.minimumSize.isFixed() && (inlineAxisSizes.maximumSize.isFixed() || inlineAxisSizes.maximumSize.isNone()));
+
+    auto& preferredSize = inlineAxisSizes.preferredSize;
+    if (auto fixedInlineSize = preferredSize.tryFixed())
+        return LayoutUnit { fixedInlineSize->resolveZoom(placedGridItem.usedZoom()) } + borderAndPadding;
+
+    if (preferredSize.isAuto()) {
+        // Grid item calculations for automatic sizes in a given dimensions vary by their
+        // self-alignment values:
+        auto alignmentPosition = placedGridItem.inlineAxisAlignment().position();
+
+        // normal:
+        // If the grid item has no preferred aspect ratio, and no natural size in the relevant
+        // axis (if it is a replaced element), the grid item is sized as for align-self: stretch.
+        //
+        // https://www.w3.org/TR/css-align-3/#propdef-align-self
+        //
+        // When the box’s computed width/height (as appropriate to the axis) is auto and neither of
+        // its margins (in the appropriate axis) are auto, sets the box’s used size to the length
+        // necessary to make its outer size as close to filling the alignment container as possible
+        // while still respecting the constraints imposed by min-height/min-width/max-height/max-width.
+        auto& marginStart = inlineAxisSizes.marginStart;
+        auto& marginEnd = inlineAxisSizes.marginEnd;
+        if ((alignmentPosition == ItemPosition::Normal) && !placedGridItem.hasPreferredAspectRatio() && !placedGridItem.isReplacedElement()
+            && !marginStart.isAuto() && !marginEnd.isAuto()) {
+            auto& usedZoom = placedGridItem.usedZoom();
+
+            auto minimumSize = LayoutUnit { inlineAxisSizes.minimumSize.tryFixed()->resolveZoom(usedZoom) };
+            auto maximumSize = [&inlineAxisSizes, &usedZoom] {
+                auto& computedMaximumSize = inlineAxisSizes.maximumSize;
+                if (computedMaximumSize.isNone())
+                    return LayoutUnit::max();
+                return LayoutUnit { computedMaximumSize.tryFixed()->resolveZoom(usedZoom) };
+            };
+
+            auto stretchedWidth = columnsSize() - LayoutUnit { marginStart.tryFixed()->resolveZoom(usedZoom) } - LayoutUnit { marginEnd.tryFixed()->resolveZoom(usedZoom) } - borderAndPadding;
```

Comment:
Can we have `resolveZoom` return a `LayoutUnit`, so we do not need to keep creating them?